src/rust/engine/fs/store/src/lib.rs

```diff
@@ -255,51 +250,32 @@ impl RemoteStore {
     };
     self
       .maybe_download(digest, async move {
-        let stored_digest = if digest.size_bytes <= IMMUTABLE_FILE_SIZE_LIMIT || f_remote.is_some()
-        {
-          // (if there's a function to call, always just buffer fully into memory)
+        let store_into_fsdb = f_remote.is_none()
+          && entry_type == EntryType::File
+          && ByteStore::should_use_fsdb(digest.size_bytes);
+        if store_into_fsdb {
+          let tempfile = local_store
+            .get_file_fsdb()
+            .get_tempfile(digest.hash)
+            .await?;
+          remote_store
+            .load_file(digest, tempfile.open().await?)
+            .await?
+            .ok_or_else(create_missing)?;
+          tempfile.persist().await?;
+        } else {
           let bytes = remote_store
             .load_bytes(digest)
             .await?
             .ok_or_else(create_missing)?;
           if let Some(f_remote) = f_remote {
             f_remote(bytes.clone())?;
           }
-          let digest = Digest::of_bytes(&bytes);
           local_store
             .store_bytes(entry_type, digest.hash, bytes, true)
             .await?;
-          digest
-        } else {
-          assert!(f_remote.is_none());
-          // TODO(#18048): choose a file that can be plopped into the local store directly, when
-          // large files are stored there
-          let file = tokio::task::spawn_blocking(tempfile::tempfile)
-            .await
-            .map_err(|e| e.to_string())??;
-          let file = tokio::fs::File::from_std(file);
-
-          let file = remote_store
-            .load_file(digest, file)
-            .await?
-            .ok_or_else(create_missing)?;
-
-          let file = file.into_std().await;
-          local_store
-            .store(entry_type, true, true, move || {
-              let mut file = file.try_clone()?;
-              file.rewind()?;
-              Ok(file)
-            })
-            .await?
-        };
-        if digest == stored_digest {
```

Comment:
Has this verification been lost? I think the bytes from the network aren't trusted, and so their contents needs to be validated against the expected digest. Maybe an explicit extra pass could be retained until  https://github.com/pantsbuild/pants/issues/18231 is worked on/lands.

Reply:
Yea, looks like we had some failing tests for this. I've re-added the validation internally to `load_monomorphic` in the latest commit, which AFAICT avoids two passes? Huzzah.